Source/Core/Common/Network.cpp

```diff
@@ -186,4 +261,204 @@ u16 ComputeNetworkChecksum(const void* data, u16 length, u32 initial_value)
     checksum = (checksum >> 16) + (checksum & 0xFFFF);
   return ~static_cast<u16>(checksum);
 }
+
+// Compute the TCP network checksum with a fake header
+u16 ComputeTCPNetworkChecksum(const IPAddress& from, const IPAddress& to, const void* data,
+                              u16 length, u8 protocol)
+{
+  // Compute the TCP checksum with its pseudo header
+  const u32 source_addr = ntohl(Common::BitCast<u32>(from));
+  const u32 destination_addr = ntohl(Common::BitCast<u32>(to));
+  const u32 initial_value = (source_addr >> 16) + (source_addr & 0xFFFF) +
+                            (destination_addr >> 16) + (destination_addr & 0xFFFF) + protocol +
+                            length;
+  const u32 tcp_checksum = ComputeNetworkChecksum(data, length, initial_value);
+  return htons(static_cast<u16>(tcp_checksum));
+}
+
+ARPPacket::ARPPacket() = default;
+
+u16 ARPPacket::Size() const
+{
+  return static_cast<u16>(SIZE);
+}
+
+TCPPacket::TCPPacket() = default;
+
+TCPPacket::TCPPacket(MACAddress destination, MACAddress source)
+{
+  eth_header.destination = destination;
+  eth_header.source = source;
+  eth_header.ethertype = htons(IPV4_ETHERTYPE);
+}
+
+std::vector<u8> TCPPacket::Build()
```

Comment:
I think there are room for improvements here. C style function (`memcpy`) should be avoided and C++ ones should be preferred. The `vector` class has member functions to insert data into it.

Reply:
using insert works, but i cant find anything to replace data in vector. i do a CRC at the end and to place it i have to erase then insert, wich is probably slow. then again according to some doc every insert is doing a new buffer so thats alot of data relocation already.

Reply:
`std::vector::insert()` only reallocates when not enough space is available in the buffer. This for example should be equivalent to the existing code, only doing a single allocation at the start:


```
std::vector<u8> TCPPacket::Build()
{
  std::vector<u8> result;
  result.reserve(Size());

  // recalc size
  ip_header.total_len = htons((u16)(IPv4Header::SIZE + ipv4_options.size() + TCPHeader::SIZE +
                                    tcp_options.size() + data.size()));

  // copy data
  result.insert(result.end(), &eth_header, &eth_header + EthernetHeader::SIZE);
  result.insert(result.end(), &ip_header, &ip_header + IPv4Header::SIZE);
  result.insert(result.end(), ipv4_options.begin(), ipv4_options.end());
  const size_t crc_offset = result.size() + offsetof(TCPHeader, checksum);
  result.insert(result.end(), &tcp_header, &tcp_header + TCPHeader::SIZE);
  result.insert(result.end(), tcp_options.begin(), tcp_options.end());
  result.insert(result.end(), data.begin(), data.end());
  tcp_header.checksum =
      ComputeTCPNetworkChecksum(ip_header.source_addr, ip_header.destination_addr,
                                &result[crc_offset], (u16)result.size() - crc_offset, IPPROTO_TCP);
  std::memcpy(&result[crc_offset], &tcp_header.checksum, sizeof(tcp_header.checksum));
  return result;
}
```

By the by, are you intentionally modifying `tcp_header.checksum` here?

Also, member variables of non-POD classes should start with an `m_`. So `m_eth_header`, `m_ip_header` and so on.

Reply:
e: ^ This doesn't actually work because `&eth_header + EthernetHeader::SIZE` doesn't actually give an end-of-object iterator equivalent... not sure how to do that without some ugly casts. Hm. But something along those lines should work, I think...

Reply:
I will also note that your checksum calculation includes the checksum field itself, but you're not taking any care of clearing it beforehand.

Reply:
Oh, one more: Instead of using `htons()` and `ntohs()` everywhere, you can just use `Common::BigEndianValue<>`, and it will automatically take care of the endian conversion.

Reply:
> e: ^ This doesn't actually work because `&eth_header + EthernetHeader::SIZE` doesn't actually give an end-of-object iterator equivalent... not sure how to do that without some ugly casts. Hm. But something along those lines should work, I think...

IIRC, I used a lambda to address that point in `NetworkCaptureLogger.cpp`. Something like this should work I believe:
```c++
  auto insert = [&](const void* obj, std::size_t size) {
    const u8* begin = reinterpret_cast<const u8*>(obj);
    result.insert(result.end(), begin, begin + size);
  };

  // copy data
  insert(&eth_header, EthernetHeader::SIZE);
  insert(&ip_header, IPv4Header::SIZE);
  ...
```

> Oh, one more: Instead of using `htons()` and `ntohs()` everywhere, you can just use `Common::BigEndianValue<>`, and it will automatically take care of the endian conversion.

I don't think that's a good idea as we don't have a `Common::NetworkEndianValue` or similar and currently `BigEndianValue` blindly swap values without caring of the host's endianness. My other fear is that it might make some of the computations less obvious if some values get implicitly converted.

Reply:
Something like this is probably a bit nicer:

```
template <typename T>
static void PushPOD(std::vector<u8>& v, const T& data)
{
  auto a = Common::BitCast<std::array<u8, sizeof(T)>, T>(data);
  v.insert(v.end(), a.begin(), a.end());
}
```

Should get optimized away, too.

Reply:
> I will also note that your checksum calculation includes the checksum field itself, but you're not taking any care of clearing it beforehand.

the initialisation does it for me but i guess it doesnt hurt to make sure